src/vmspawn/vmspawn.c

```diff
@@ -390,7 +638,38 @@ static int run_virtual_machine(void) {
                 _exit(EXIT_FAILURE);
         }
 
-        return wait_for_terminate_and_check(qemu_binary, child_pid, WAIT_LOG);
+#ifdef ARCHITECTURE_SUPPORTS_SMBIOS
+        int exit_status = INT_MAX;
+        if (use_vsock)
+                setup_notify_parent(event, vsock_fd, &exit_status, &notify_event_source);
+#endif
+
+        /* shutdown qemu when we are shutdown */
+        (void) sd_event_add_signal(event, NULL, SIGINT, on_orderly_shutdown, PID_TO_PTR(child_pid));
+        (void) sd_event_add_signal(event, NULL, SIGTERM, on_orderly_shutdown, PID_TO_PTR(child_pid));
+
+        (void) sd_event_add_signal(event, NULL, SIGRTMIN+18, sigrtmin18_handler, NULL);
+
+        /* Exit when the child exits */
+        (void) sd_event_add_child(event, NULL, child_pid, WEXITED, on_child_exit, NULL);
+
+        r = sd_event_loop(event);
+        if (r < 0)
+                return log_error_errno(r, "Failed to run event loop: %m");
+
+#ifdef ARCHITECTURE_SUPPORTS_SMBIOS
+        if (use_vsock) {
+                if (exit_status == INT_MAX) {
+                        log_warning("Couldn't retrieve inner EXIT_STATUS from vsock");
+                        return EXIT_FAILURE;
+                }
```

Comment:
whether a VM actually hands this data to us or not should really be optional. if it never sends an EXIT_STATUS to us this should be logged at debug level but not more, and not result in EXIT_FAILURE.

Reply:
Fixed in 7126b0eb12.